File: atcoder.jp/contests/abc200/tasks/abc200_b/main.ts
import { readFileSync } from "fs";
let inputLinesIndex = 0;
const inputLines = readFileSync("/dev/stdin", "utf8").split("\n");
const input = () => inputLines[inputLinesIndex++];

function* iterate<T>(func: (arg: T) => T, start: T) {
  while (true) {
    yield start;
    start = func(start);
  }
}

const nth = <T>(iterable: Generator<T, void, unknown>, n: number) => {
  for (let i = 0; i < n; i++) iterable.next();
  return iterable.next().value;
};

const func = (n: number) => (n % 200 === 0 ? n / 200 : 1000 * n + 200);

const main = () => {
  const [N, K] = input().split(" ").map(Number);

  const ans = nth(iterate(func, N), K);
  console.log(ans);
};

main();

File: atcoder.jp/contests/abc200/tasks/abc200_b/main.py
from itertools import islice
from typing import Callable, Iterable, Iterator, TypeVar, Union

_T = TypeVar("_T")
_U = TypeVar("_U")


# more-itertools
def iterate(func: Callable[[_T], _T], start: _T) -> Iterator[_T]:
    # https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.iterate
    while True:
        yield start
        start = func(start)


def nth(iterable: Iterable[_T], n: int, default: _U = None) -> Union[_T, _U]:
    # https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.nth
    return next(islice(iterable, n, None), default)


def func(n: int) -> int:
    if n % 200 == 0:
        return n // 200
    return 1000 * n + 200


N, K = map(int, input().split())

ans = nth(iterate(func, N), K)
print(ans)
